### ISSUE TITLE
Single page for all text or a search function [request]

### ISSUE THREAD

**Issue by 678GUY** [felikatze/isat-script-project#199]
i think having a page where all the text is in one page would let people search for text, or at least some form of basic search function to be able to find stuff

**Comment by goldstargloww** [comment]
i think a single page for all text would be a bit of a hassle both to maintain and use, but i can look into figuring out how a search function across all pages would work

**Comment by 678GUY** [comment]
> i think a single page for all text would be a bit of a hassle both to maintain and use, but i can look into figuring out how a search function across all pages would work

you could assign each page an order number and the all text page would pull all text by the order so that it wouldnt need active maintaining, however i dont know how the site is built and works so my suggestion might be useless, either way thank you!

**Comment by goldstargloww** [comment]
good suggestion on the pulling from other pages automatically actually, though i think it'd be pretty difficult to get looking right, and it's worth noting that it would take a while to load lol

either way we'll think about it and i'll look into how search would work :>

**Comment by 678GUY** [comment]
I think that looking right matters less when searching, at worst each page can be linked, but either way I just think being able to find stuff is what matters regardless of the method, so thank you very much for considering it (:

**Comment by raspberryspider** [comment]
i agree with gold, i think making one page with all our text would be a major pain to maintain manually. by the looks of things it's not super feasible to have text mirror automatically either (mainly the solutions i saw are using PHP, which neocities doesn't natively support. we'd need another server host, of which there doesn't seem to be any not-dodgy free ones).

as for a search bar, for much the same reasons as the text mirroring, there doesn't seem to be a great way to make our own search bar without hosting backend elsewhere and/or exploding our bandwidth. we could implement a custom search engine shenanigan(??), so our options are:
- [google custom programmable search engine](https://developers.google.com/custom-search/docs/tutorial/creatingcse). don't love this one tbh it just looks inelegant. big ass POWERED BY GOOGLE flashing strobe lights the whole nine yards. the upside is that i believe we can implement it onsite rather than needing to redirect
- the other big option is using a search box that sends users to another search engine with custom parameters that search just the site. i believe you can do this with ecosia but i cannot for the life of me find their docs. i found [duckduckgo's search URL params documentation](https://duckduckgo.com/duckduckgo-help-pages/settings/params/) though, and also [someone who wrote the html](https://ddg.patdryburgh.com/) for a form that makes the custom url (take the \#s out of the colors tho it doesn't work). downside is we have to redirect users out to the search engine's site, and i couldn't get "open in a new tab(/window?)" to work
- relatedly, ddg does seem to have us indexed appropiately regularly? here's a search for [spoilers](https://duckduckgo.com/?sites=felicitations.neocities.org&k7=000000&k8=ffffff&k9=g&kx=e&kj=434343&kt=a&q=spoilers&atb=v308-1&ia=web) that grabbed mira's act 6 end talk from two-ish days ago (yes the billion params are supposed to be there it is one of my biggest pet peeves when people don't clean urls they're sharing i would not do this with no reason)

**Comment by goldstargloww** [comment]
> i think making one page with all our text would be a major pain to maintain manually. by the looks of things it's not super feasible to have text mirror automatically either (mainly the solutions i saw are using PHP, which neocities doesn't natively support. we'd need another server host, of which there doesn't seem to be any not-dodgy free ones).

i haven't looked into it, but i was thinking javascript might be able to do it? i know it can at least insert entire stretches of a document, but i don't know about grabbing text from other documents

another thought is to use iframes or otherwise embedded pages within the page, so it'd just be like, a page with every other page embedded inside of it (though this might need to be updated manually to add new pages, though it shouldn't be too painful; not too much on top of the updating rss + overview pages). though this does have me thinking about dialogue options and `<details>`, which you can't ctrl+f inside of without opening first, and i'm not sure if the toggle dialogue options button on the main page would work for the embedded pages as well

> * [google custom programmable search engine](https://developers.google.com/custom-search/docs/tutorial/creatingcse). don't love this one tbh it just looks inelegant. big ass POWERED BY GOOGLE flashing strobe lights the whole nine yards. the upside is that i believe we can implement it onsite rather than needing to redirect

yeah, it looks like this is what w3schools uses for [its own search](https://www.w3schools.com/#gsc.tab=0&gsc.q=query), and it's definitely jarring and out of place. though it does look like we *might* be able to mess with its css? but that sounds painful

![image](https://github.com/user-attachments/assets/8b1ac5b0-5e40-4984-86ec-0a8ab58352df)


> * the other big option is using a search box that sends users to another search engine with custom parameters that search just the site. i believe you can do this with ecosia but i cannot for the life of me find their docs. i found [duckduckgo's search URL params documentation](https://duckduckgo.com/duckduckgo-help-pages/settings/params/) though, and also [someone who wrote the html](https://ddg.patdryburgh.com/) for a form that makes the custom url (take the #s out of the colors tho it doesn't work). downside is we have to redirect users out to the search engine's site, and i couldn't get "open in a new tab(/window?)" to work

yeahh not ideal but could work i guess

> * relatedly, ddg does seem to have us indexed appropiately regularly? here's a search for [spoilers](https://duckduckgo.com/?sites=felicitations.neocities.org&k7=000000&k8=ffffff&k9=g&kx=e&kj=434343&kt=a&q=spoilers&atb=v308-1&ia=web) that grabbed mira's act 6 end talk from two-ish days ago (yes the billion params are supposed to be there it is one of my biggest pet peeves when people don't clean urls they're sharing i would not do this with no reason)

you can do [the same with google](https://www.google.com/search?q=spoilers+site%3Afelicitations.neocities.org), i think most search engines use a pretty similar sort of format for that—both google and [ecosia](https://www.ecosia.org/search?q=spoilers+site%3Afelicitations.neocities.org) seem to give the same if not similar results, both of which are more than what duckduckgo gives, but duckduckgo seems to only give the one where the word "spoilers" is actually in dialogue for some reason? might just be a coincidence, shrug

also :O someone who also hates unnecessary url parameters :D

**Comment by goldstargloww** [comment]
@LunarDusk1 nobody's started, no—if you wanna try figuring out something go ahead!

**Comment by LunarDusk1** [comment]
Minimalist version of a search bar. Since there's effectively no backend, a full search requires to query every page containing dialogue once, then storing it locally for processing. I did a few things to avoid unnecessary requests (only get the data once the user has clicked on the search bar, etc). I think this is probably the most efficient way of doing things in the current setup.

If bandwidth is an issue, there could be other options to store dialogues in a separate file and process that, but I wanted to start by making a version that requires the fewest number of changes to the current setup.

In any case, still lots more work to do, if there's any comments/recommendations on how to make this fit the project better please let me know! Will publish the fork sometime tomorrow.

https://github.com/user-attachments/assets/f7d79eb0-efd9-41be-b6fb-0077e4c22495

**Comment by felikatze** [comment]
@LunarDusk1 That is super impressive wow. About backend: since we're hosted on neocities, we do not have access to any server-side scripting whatsoever. We also have a alotted bandwith of 200 GB, which I am currently not all too worried about, since the entire site sits at a whopping 13.35 MB atm. We also only really discussed having everything on one page for search purposes as an end term goal when we're actually done with the script, due to the maintanance issues mentioned in the discussion previously.

The only thing what you showed would really need to fit in would be to style the search bar visually. Maybe as a separate strip below the header bar?

Either way, wow, and thank you so much for looking into this for us!

**Comment by LunarDusk1** [comment]
@felikatze Thank you for the kind words! What you mentioned about this being an end-term goal makes sense, even with the way I'm doing this, since the pages do need to be listed manually with this method and so the list would need to be expanded as more pages are added. I do still think this is a nice feature to have as the website is growing, even if it's not entirely accurate to everything available (although of course the content added could be periodically added to the search dataset).

### Anyways, here's an update on the **progress of this feature**, including:
- Visuals changed to be more in-line with the rest of the website
- Search bar is now in the header, across all pages (as far as I know, more info in the to-be-improved section)
- Search results doesn't push down the rest of the elements lol
- Little "..." to show the loading of the data
- The list showing the search results can appear/disappear depending on where the user clicks on the page
- Fixed various bugs where text wouldn't show up if it was wavey, big, wish-craft-ey, and about 10000 little things

I'm pretty happy with the current product, enough that I've used it on my own to look up specific lines lol. 

### With that said, there are still some **important improvements to be made**:
- The code, especially the CSS I added, is currently a bit of a mess. I need to clean it up, make sure everything's in its place, etc.
- Performance should be improved on the loading of the data, especially important for the actual neocities page.
- Search results should be labeled with which page they send to
- I'm not 100% sure every page has the searchbar and an intact header, since I used text-replace to add the new script tag to the 140+ different pages. Need to make sure they're all alright (possibly by creating a universal head tag).
- There's probably a better place to put a searchbar like this, it feels like it takes up a bit too much of the header. Gotta test different positions.

Like I said, I'm happy with the progress made, though I personally wouldn't put it up until more extensive testing has been done. Will push to my fork in a little bit. Is there a way I could directly contact someone who's worked on this project for a long time and is familiar with how the different pages are structured? I'd need info to help sort through dialogue more effectively.

https://github.com/user-attachments/assets/28da43e5-93c2-4180-9416-dc3f8481256c

**Comment by goldstargloww** [comment]
@LunarDusk1 you could talk to me on discord (@goldstargloww) if you want and could probably join the script project discord as well :>
